File: Yarn/session.go
package yarn

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"
	"sync"
	"time"

	"github.com/google/uuid"
)

// Session is a named research session grouping conversations and measurements.
type Session struct {
	ID          string         `json:"id"`
	Name        string         `json:"name"`
	Description string         `json:"description"`
	StartedAt   time.Time      `json:"started_at"`
	EndedAt     *time.Time     `json:"ended_at,omitempty"`
	Config      SessionConfig  `json:"config"`
	Metadata    map[string]any `json:"metadata,omitempty"`

	Conversations []*Conversation `json:"conversations"`
	Measurements  []*Measurement  `json:"measurements"`

	mu sync.RWMutex
}

// SessionConfig holds session configuration.
type SessionConfig struct {
	MeasurementMode MeasurementMode `json:"measurement_mode"`
	AutoExport      bool            `json:"auto_export"`
	ExportPath      string          `json:"export_path"`
}

// MeasurementMode determines when measurements are captured.
type MeasurementMode string

const (
	MeasurePassive   MeasurementMode = "passive"   // Observe only
	MeasureActive    MeasurementMode = "active"    // Every exchange
	MeasureTriggered MeasurementMode = "triggered" // On request
)

// IsValid returns true if this is a valid measurement mode.
func (m MeasurementMode) IsValid() bool {
	switch m {
	case MeasurePassive, MeasureActive, MeasureTriggered:
		return true
	default:
		return false
	}
}

// Validate checks if the session config is valid.
// Returns a ValidationError if invalid, nil if valid.
// Empty MeasurementMode is allowed (treated as default).
func (c *SessionConfig) Validate() *ValidationError {
	// MeasurementMode must be valid if set
	if c.MeasurementMode != "" && !c.MeasurementMode.IsValid() {
		return &ValidationError{Field: "measurement_mode", Message: "invalid measurement mode"}
	}
	return nil
}

// Validate checks if the session is valid.
// Returns a ValidationError if invalid, nil if valid.
func (s *Session) Validate() *ValidationError {
	if s.ID == "" {
		return &ValidationError{Field: "id", Message: "id is required"}
	}
	if s.Name == "" {
		return &ValidationError{Field: "name", Message: "name is required"}
	}
	if s.StartedAt.IsZero() {
		return &ValidationError{Field: "started_at", Message: "started_at is required"}
	}

	// EndedAt must be after StartedAt if set
	if s.EndedAt != nil && !s.EndedAt.IsZero() {
		if s.EndedAt.Before(s.StartedAt) || s.EndedAt.Equal(s.StartedAt) {
			return &ValidationError{Field: "ended_at", Message: "ended_at must be after started_at"}
		}
	}

	// Validate MeasurementMode if set
	if s.Config.MeasurementMode != "" && !s.Config.MeasurementMode.IsValid() {
		return &ValidationError{Field: "config.measurement_mode", Message: "invalid measurement mode"}
	}

	// Cascade validation: validate all conversations
	for i, conv := range s.Conversations {
		if conv == nil {
			return &ValidationError{
				Field:   "conversations",
				Message: "conversation at index " + strconv.Itoa(i) + " is nil",
			}
		}
		if err := conv.Validate(); err != nil {
			return &ValidationError{
				Field:   "conversations[" + strconv.Itoa(i) + "]." + err.Field,
				Message: err.Message,
			}
		}
	}

	// Cascade validation: validate all measurements
	for i, m := range s.Measurements {
		if m == nil {
			return &ValidationError{
				Field:   "measurements",
				Message: "measurement at index " + strconv.Itoa(i) + " is nil",
			}
		}
		if err := m.Validate(); err != nil {
			return &ValidationError{
				Field:   "measurements[" + strconv.Itoa(i) + "]." + err.Field,
				Message: err.Message,
			}
		}
	}

	return nil
}

// NewSession creates a new session.
func NewSession(name, description string) *Session {
	return &Session{
		ID:            uuid.New().String(),
		Name:          name,
		Description:   description,
		StartedAt:     time.Now(),
		Conversations: make([]*Conversation, 0),
		Measurements:  make([]*Measurement, 0),
		Metadata:      make(map[string]any),
		Config: SessionConfig{
			MeasurementMode: MeasureActive,
			AutoExport:      true,
			ExportPath:      "./experiments",
		},
	}
}

// AddConversation adds a conversation to the session.
func (s *Session) AddConversation(conv *Conversation) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.Conversations = append(s.Conversations, conv)
}

// AddMeasurement adds a measurement to the session.
func (s *Session) AddMeasurement(m *Measurement) {
	s.mu.Lock()
	defer s.mu.Unlock()
	m.SessionID = s.ID
	s.Measurements = append(s.Measurements, m)
}

// ActiveConversation returns the most recent conversation, or creates one.
func (s *Session) ActiveConversation() *Conversation {
	s.mu.Lock()
	defer s.mu.Unlock()

	if len(s.Conversations) == 0 {
		conv := NewConversation(s.Name + "-conv-1")
		s.Conversations = append(s.Conversations, conv)
	}
	return s.Conversations[len(s.Conversations)-1]
}

// End marks the session as ended.
func (s *Session) End() {
	s.mu.Lock()
	defer s.mu.Unlock()
	now := time.Now()
	s.EndedAt = &now
}

// Stats returns session statistics.
func (s *Session) Stats() SessionStats {
	s.mu.RLock()
	defer s.mu.RUnlock()

	stats := SessionStats{
		ConversationCount: len(s.Conversations),
		MeasurementCount:  len(s.Measurements),
	}

	for _, conv := range s.Conversations {
		stats.MessageCount += conv.Length()
	}

	if len(s.Measurements) > 0 {
		var totalDEff, totalBeta, totalAlignment float64
		var bilateralCount int

		for _, m := range s.Measurements {
			totalDEff += float64(m.DEff)
			totalBeta += m.Beta
			totalAlignment += m.Alignment
			if m.IsBilateral() {
				bilateralCount++
			}
		}

		n := float64(len(s.Measurements))
		stats.AvgDEff = totalDEff / n
		stats.AvgBeta = totalBeta / n
		stats.AvgAlignment = totalAlignment / n
		stats.BilateralCount = bilateralCount
	}

	return stats
}

// SessionStats holds session statistics.
type SessionStats struct {
	ConversationCount int     `json:"conversation_count"`
	MessageCount      int     `json:"message_count"`
	MeasurementCount  int     `json:"measurement_count"`
	BilateralCount    int     `json:"bilateral_count"`
	AvgDEff           float64 `json:"avg_d_eff"`
	AvgBeta           float64 `json:"avg_beta"`
	AvgAlignment      float64 `json:"avg_alignment"`
}

// Export writes the session to files (JSON + JSONL for measurements).
func (s *Session) Export() (err error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	exportDir := filepath.Join(s.Config.ExportPath, s.ID)
	if err := os.MkdirAll(exportDir, 0755); err != nil {
		return err
	}

	// Export session metadata
	sessionFile := filepath.Join(exportDir, "session.json")
	sessionData, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return err
	}
	if err := os.WriteFile(sessionFile, sessionData, 0644); err != nil {
		return err
	}

	// Export measurements as JSONL
	measurementsFile := filepath.Join(exportDir, "measurements.jsonl")
	f, err := os.Create(measurementsFile)
	if err != nil {
		return err
	}
	defer func() {
		if cerr := f.Close(); cerr != nil && err == nil {
			err = cerr
		}
	}()

	for _, m := range s.Measurements {
		data, err := json.Marshal(m)
		if err != nil {
			return err
		}
		if _, err := f.Write(data); err != nil {
			return err
		}
		if _, err := f.WriteString("\n"); err != nil {
			return err
		}
	}

	return nil
}

File: web-ui/src/types/session.ts
/**
 * Session types matching Yarn/session.go and Yarn/conversation.go
 * Research session management types.
 */

import type { Measurement } from './measurement';
import type { HiddenState } from './backend';

/** MessageRole represents the sender type. */
export type MessageRole = 'system' | 'user' | 'assistant' | 'tool';

/** Check if a role is valid. */
export function isValidMessageRole(role: string): role is MessageRole {
  return ['system', 'user', 'assistant', 'tool'].includes(role);
}

/** Message is the atomic unit of communication between agents. */
export interface Message {
  id: string;
  role: MessageRole;
  content: string;
  timestamp: string; // ISO 8601 date string
  agent_id?: string;
  agent_name?: string;
  hidden_state?: HiddenState;
  metadata?: Record<string, unknown>;

  // Tool-related fields
  tool_call_id?: string;
  tool_name?: string;
}

/** Participant tracks an agent's involvement in the conversation. */
export interface Participant {
  agent_id: string;
  agent_name: string;
  role: string;
  joined_at: string; // ISO 8601 date string
  message_count: number;
}

/** Conversation is an ordered sequence of messages with participant tracking. */
export interface Conversation {
  id: string;
  name: string;
  messages: Message[];
  participants: Record<string, Participant>;
  created_at: string; // ISO 8601 date string
  updated_at: string; // ISO 8601 date string
  metadata?: Record<string, unknown>;
}

/** MeasurementMode determines when measurements are captured. */
export type MeasurementMode = 'passive' | 'active' | 'triggered';

/** Check if a measurement mode is valid. */
export function isValidMeasurementMode(mode: string): mode is MeasurementMode {
  return ['passive', 'active', 'triggered'].includes(mode);
}

/** SessionConfig holds session configuration (runtime). */
export interface SessionConfig {
  measurement_mode: MeasurementMode;
  auto_export: boolean;
  export_path: string;
}

/** Session is a named research session grouping conversations and measurements. */
export interface Session {
  id: string;
  name: string;
  description: string;
  started_at: string; // ISO 8601 date string
  ended_at?: string; // ISO 8601 date string
  config: SessionConfig;
  metadata?: Record<string, unknown>;
  conversations: Conversation[];
  measurements: Measurement[];
}

/** SessionStats holds session statistics. */
export interface SessionStats {
  conversation_count: number;
  message_count: number;
  measurement_count: number;
  bilateral_count: number;
  avg_d_eff: number;
  avg_beta: number;
  avg_alignment: number;
}

/** SessionStatus represents the current status of a session in the registry. */
export interface SessionStatus {
  name: string;
  id: string;
  is_active: boolean;
  started_at: string; // ISO 8601 date string
  ended_at?: string; // ISO 8601 date string
  stats: SessionStats;
}

/** Calculate session statistics from session data. */
export function calculateSessionStats(session: Session): SessionStats {
  const stats: SessionStats = {
    conversation_count: session.conversations.length,
    message_count: 0,
    measurement_count: session.measurements.length,
    bilateral_count: 0,
    avg_d_eff: 0,
    avg_beta: 0,
    avg_alignment: 0,
  };

  // Count messages across all conversations
  for (const conv of session.conversations) {
    stats.message_count += conv.messages.length;
  }

  // Calculate averages from measurements
  if (session.measurements.length > 0) {
    let totalDEff = 0;
    let totalBeta = 0;
    let totalAlignment = 0;

    for (const m of session.measurements) {
      totalDEff += m.d_eff;
      totalBeta += m.beta;
      totalAlignment += m.alignment;

      // Count bilateral measurements
      const hasSender = m.sender_hidden && m.sender_hidden.vector.length > 0;
      const hasReceiver = m.receiver_hidden && m.receiver_hidden.vector.length > 0;
      if (hasSender && hasReceiver) {
        stats.bilateral_count++;
      }
    }

    const n = session.measurements.length;
    stats.avg_d_eff = totalDEff / n;
    stats.avg_beta = totalBeta / n;
    stats.avg_alignment = totalAlignment / n;
  }

  return stats;
}
